lib/tbot/workloadidentity/workloadattest/sigstore.go

```diff
@@ -0,0 +1,166 @@
+		if err != nil {
+			return trace.Wrap(err, "checking credentials_path")
+		}
+		if fi.IsDir() {
+			return trace.BadParameter("credentials_path cannot be a directory")
+		}
+	}
+
+	for idx, reg := range s.AdditionalRegistries {
+		field := fmt.Sprintf("additional_registries[%d].host", idx)
+
+		if reg.Host == "" {
+			return trace.BadParameter("%s cannot be blank", field)
+		}
+
+		if _, err := name.NewRegistry(reg.Host); err != nil {
+			return trace.Wrap(err, "parsing %s", field)
+		}
+	}
+
+	return nil
+}
+
+// SigstoreRegistryConfig holds configuration for an OCI registry the Sigstore
+// attestor will check for signatures and attestations.
+type SigstoreRegistryConfig struct {
+	// Host is the hostname (and optionally the port) of the registry.
+	Host string `yaml:"host"`
+}
+
+// SigstoreAttestor discovers signatures and attestations for a container image.
+type SigstoreAttestor struct {
+	cfg SigstoreAttestorConfig
+	log *slog.Logger
+
+	registryHosts []string
+
+	keychain authn.Keychain
+	cache    *lru.Cache[string, *workloadidentityv1.WorkloadAttrsSigstore]
+}
+
+// NewSigstoreAttestor creates a new SigstoreAttestor with the given configuration.
+func NewSigstoreAttestor(cfg SigstoreAttestorConfig, log *slog.Logger) (*SigstoreAttestor, error) {
+	keychain, err := sigstore.Keychain(cfg.CredentialsPath)
+	if err != nil {
+		return nil, trace.Wrap(err, "loading credentials")
+	}
+
+	cache, err := lru.New[string, *workloadidentityv1.WorkloadAttrsSigstore](64)
+	if err != nil {
+		return nil, trace.Wrap(err, "building LRU cache")
+	}
+
+	regHosts := make([]string, len(cfg.AdditionalRegistries))
+	for idx, reg := range cfg.AdditionalRegistries {
+		regHosts[idx] = reg.Host
+	}
+
+	return &SigstoreAttestor{
+		cfg:           cfg,
+		log:           log,
+		registryHosts: regHosts,
+		keychain:      keychain,
+		cache:         cache,
+	}, nil
+}
+
+// Attest discovers signatures and attestations for the given container.
+func (a *SigstoreAttestor) Attest(ctx context.Context, ctr Container) (*workloadidentityv1.WorkloadAttrsSigstore, error) {
+	logger := a.log.With(
+		"image", ctr.GetImage(),
+		"image_digest", ctr.GetImageDigest(),
+	)
+	logger.InfoContext(ctx, "Starting Sigstore workload attestation")
+
+	if cached, ok := a.cache.Get(ctr.GetImageDigest()); ok {
+		logger.DebugContext(ctx, "Sigstore attestor cache hit")
+		return cached, nil
+	}
+	logger.DebugContext(ctx, "Sigstore attestor cache miss")
+
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+
+	payloads, err := sigstore.Discover(ctx,
+		ctr.GetImage(),
+		ctr.GetImageDigest(),
+		sigstore.DiscoveryConfig{
+			Logger:               a.log,
+			Keychain:             a.keychain,
+			AdditionalRegistries: a.registryHosts,
+		},
+	)
+	if err != nil {
+		return nil, trace.Wrap(err, "discovering Sigstore payloads")
+	}
+
+	result := &workloadidentityv1.WorkloadAttrsSigstore{
+		Payloads: payloads,
+	}
+	_ = a.cache.Add(ctr.GetImageDigest(), result)
```

Comment:
I wonder if we should avoid caching an empty payloads list in case the image hasn't been signed _yet_? Or maybe we should use a TTL-based cache rather than a simple LRU.

Reply:
Hm - perhaps TTL does make more sense here - or TTL w/ LRU. If anything, I'd worry that a workload that fails to get an identity is likely to enter some kind of crashloop and we'd end up hitting the upstream registry quite a lot if we don't have any form of caching.

Reply:
Thinking about this a bit more, I'd really like to keep the property that if we can continue to get a workload identity using the signatures we've discovered, we never (need to) refresh them. So a standard TTL-based cache doesn't quite fit.

Instead, in fc68549e5596f840e50a0bf1e6822e24aae1313a I've added a supplementary hash where we track the time at which using the signatures failed. If using the signatures fails, we'll refresh them after 30s (to prevent hammering registries) otherwise we'll just keep using the cached value in perpetuity.